src/SerialIO.c

```diff
@@ -0,0 +1,58 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "SerialIO.h"
+#include <stdio.h>
+#include "Serial.h"
+
+static int
+Print(uint32_t aLength, void* aBuffer)
+{
+  IPCMessageQueue* queue = GetSerialOutQueue();
+  if (!queue) {
+    return -1;
+  }
+
+  IPCMessage msg;
+  int res = IPCMessageInit(&msg);
+  if (res < 0) {
+    return -1;
+  }
+  res = IPCMessageProduce(&msg, aLength, aBuffer);
+  if (res < 0) {
+    return -1;
+  }
+  res = IPCMessageQueueConsume(queue, &msg);
+  if (res < 0) {
+    return -1;
+  }
+  return IPCMessageWaitForConsumption(&msg);
+}
+
+/*
+ * Libc-like interfaces for easy usage.
+ */
+
+int
+print(const char* fmt, ...)
+{
+  va_list ap;
+
+  va_start(ap, fmt);
+  int res = vprint(fmt, ap);
+  va_end(ap);
+
+  return res;
+}
+
+int
+vprint(const char* fmt, va_list ap)
+{
+  char buf[128];
+  int res = vsnprintf(buf, sizeof(buf), fmt, ap);
+  if (res < 0) {
+    return -1;
+  }
+  return Print(sizeof(buf), buf);
```

Comment:
I also have a implementation of printf over here:
https://github.com/dhylands/projects/blob/master/common/StrPrintf.c#L189
which takes a pointer to a function that outputs a single char. This version doesn't have floating point support, so its quite a bit smaller that the regular sprintf code. Because it takes a pointer to a function you don't need to allocate stack buffers to hold the intermediate results.